### ISSUE TITLE
Itakura parallelogram definition with time series of different lengths

### ISSUE THREAD

**Issue by rtavenar** [rtavenar/tslearn#140]
I am experimenting with `tslearn` v0.2 and I have the following issue:

```python

In [1]: import tslearn

In [2]: print(tslearn.__version__)                                                                                                                                                                                                  
0.2.0

In [3]: from tslearn.metrics import dtw                                                                                                                                                                                             

In [4]: import numpy as np                                                                                                                                                                                                          

In [5]: ts1 = np.zeros((100, 1))                                                                                                                                                                                                    

In [6]: ts2 = np.zeros((10, 1))                                                                                                                                                                                                     

In [7]: dtw(ts1, ts2, global_constraint="sakoe_chiba", sakoe_chiba_radius=1)                                                                                                                                                        
Out[7]: inf
```

This is due to the definition of the Sakoe-Chiba band as it is at the moment: if time series have different length, it might be that the term at index `(n, m)` (where n and m are the series lengths) is outside the band, which doesn't make sense.

I think we should fix it asap.

My suggestion for the fix is to consider, for a given row `i`, the element at index `i * sz2 / sz1` (or somethig like that) as the median element of the band.

**Comment by rtavenar** [comment]
@johannfaouzi I can take care of that PR, but I would be interested in your feedback about how to properly deal with it.

**Comment by rtavenar** [comment]
A bit of insight. In fact, I was wrong, the problem is not at the end index:

```python
In [13]: from tslearn.metrics import itakura_mask                                                                                                                                                                                   

In [14]: itakura_mask(10, 3)                                                                                                                                                                                                        
Out[14]: 
array([[ 0., inf, inf],
       [inf, inf, inf],
       [inf, inf, inf],
       [inf,  0., inf],
       [inf,  0., inf],
       [inf,  0., inf],
       [inf,  0., inf],
       [inf, inf, inf],
       [inf, inf, inf],
       [inf, inf,  0.]])
```

I still believe that the notion of median index per row would make sense to avoid such cases.

**Comment by rtavenar** [comment]
OK I was confusing Itakura and SakoeChiba in my previous post.

And the surprising bit is that:

```python
In [1]: from tslearn.metrics import sakoe_chiba_mask                                                                                                                                                                                

In [2]: sakoe_chiba_mask(10, 3, 1)                                                                                                                                                                                                  
Out[2]: 
array([[ 0.,  0., inf],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [inf,  0.,  0.]])

In [3]: from tslearn.metrics import dtw                                                                                                                                                                                             

In [4]: import numpy as np                                                                                                                                                                                                 

In [5]: ts1 = np.zeros((10, 1))                                                                                                                                                                                                     

In [6]: ts2 = np.zeros((3, 1))                                                                                                                                                                                                      

In [7]: dtw(ts1, ts2, global_constraint="sakoe_chiba", sakoe_chiba_radius=1)                                                                                                                                                        
Out[7]: inf
```

So I'm a bit confused at the moment :/

**Comment by rtavenar** [comment]
OK, spotted 2 bugs in fact:

1. Sakoe-Chiba and Itakura masks are permuted in `compute_mask`
2. Itakura parallelogram has an issue when sizes differ a lot

So the first one is easy to fix, but the second one requires to dig a bit more.

**Comment by rtavenar** [comment]
Fixed in version 0.2.1 (available next monday on PyPI, right now on master).

**Comment by johannfaouzi** [comment]
I looked a bit into it and the issue occurs when `max_slope` is lower than `max(sz1, sz2) / min (sz1, sz2)`.

When `sz1 == sz2`, one wants `max_slope` to be larger than 1. When the sizes are different, `max_slope` must be large enough (greater than `max(sz1, sz2) / min (sz1, sz2)`) so that there is at least one acceptable path.

We could also rescale the `max_slope` parameter so that it takes into account the difference of sizes. Adding this line at the beginning of the computation of the mask (first line of `_njit_itakura_mask`):
```python
max_slope = float(max_slope) * max(sz1, sz2) / float(min(sz1, sz2))
```
does the trick, and the parallelogram is always well defined (as long as `max_slope` is greater than or equal to 1), regardless of the sizes.
